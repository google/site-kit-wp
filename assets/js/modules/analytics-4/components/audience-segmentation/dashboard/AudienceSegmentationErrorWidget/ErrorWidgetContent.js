/**
 * ErrorWidgetContent component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { forwardRef } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { Button } from 'googlesitekit-components';
import { Cell, Grid, Row } from '../../../../../../material-components';
import {
	BREAKPOINT_SMALL,
	BREAKPOINT_TABLET,
	useBreakpoint,
} from '../../../../../../hooks/useBreakpoint';
import AudienceSegmentationErrorSVG from '../../../../../../../svg/graphics/audience-segmentation-error-full-width.svg';
import { isInsufficientPermissionsError } from '../../../../../../util/errors';
import ReportErrorActions from '../../../../../../components/ReportErrorActions';
import GetHelpLink from '../GetHelpLink';

const ErrorWidgetContent = forwardRef(
	( { Widget, errors, onRetry, onRequestAccess, showRetryButton }, ref ) => {
		const breakpoint = useBreakpoint();
		const isMobileBreakpoint = breakpoint === BREAKPOINT_SMALL;
		const isTabletBreakpoint = breakpoint === BREAKPOINT_TABLET;

		const hasInsufficientPermissionsError = errors.some(
			isInsufficientPermissionsError
		);

		return (
			<Widget
				ref={ ref }
				noPadding
				className="googlesitekit-audience-segmentation-error-widget"
			>
				<Grid
					collapsed
					className="googlesitekit-audience-segmentation-error__widget-primary-cell"
				>
					<Row>
						<Cell smSize={ 6 } mdSize={ 8 } lgSize={ 7 }>
							<h3 className="googlesitekit-publisher-win__title">
								{ hasInsufficientPermissionsError
									? __(
											'Insufficient permissions',
											'google-site-kit'
									  )
									: __(
											'Your visitor groups data loading failed',
											'google-site-kit'
									  ) }
							</h3>
							<div className="googlesitekit-widget-audience-segmentation-error__actions">
								{ showRetryButton && onRetry ? (
									<Button onClick={ onRetry } danger>
										{ __( 'Retry', 'google-site-kit' ) }
									</Button>
								) : (
									<ReportErrorActions
										moduleSlug="analytics-4"
										error={ errors }
										GetHelpLink={
											hasInsufficientPermissionsError
												? GetHelpLink
												: undefined
										}
										hideGetHelpLink={
											! hasInsufficientPermissionsError
										}
										buttonVariant="danger"
										getHelpClassName="googlesitekit-error-retry-text"
										onRetry={ onRetry }
										onRequestAccess={ onRequestAccess }
									/>
								) }
							</div>
						</Cell>
						{ ! isMobileBreakpoint && ! isTabletBreakpoint && (
							<Cell
								className="googlesitekit-widget-audience-segmentation-error__svg-wrapper"
								smSize={ 6 }
								mdSize={ 3 }
								lgSize={ 5 }
							>
								<AudienceSegmentationErrorSVG width="233px" />
							</Cell>
						) }
						{ isTabletBreakpoint && (
							<Cell
								className="googlesitekit-widget-audience-segmentation-error__svg-wrapper"
								mdSize={ 8 }
							>
								<AudienceSegmentationErrorSVG width="233px" />
							</Cell>
						) }
						{ isMobileBreakpoint && (
							<Cell
								className="googlesitekit-widget-audience-segmentation-error__svg-wrapper"
								smSize={ 8 }
							>
								<AudienceSegmentationErrorSVG width="233px" />
							</Cell>
						) }
					</Row>
				</Grid>
			</Widget>
		);
	}
);

ErrorWidgetContent.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	errors: PropTypes.arrayOf( PropTypes.object ).isRequired,
	onRetry: PropTypes.func.isRequired,
	onRequestAccess: PropTypes.func.isRequired,
	showRetryButton: PropTypes.bool,
};

export default ErrorWidgetContent;
