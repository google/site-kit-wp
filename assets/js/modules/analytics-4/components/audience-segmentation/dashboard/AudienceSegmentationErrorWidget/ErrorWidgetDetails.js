/**
 * ErrorWidgetDetails component.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import P from '@/js/components/Typography/P';
import { Button } from 'googlesitekit-components';
import ReportErrorActions from '@/js/components/ReportErrorActions';
import GetHelpLink from '@/js/modules/analytics-4/components/audience-segmentation/dashboard/GetHelpLink';

export default function ErrorWidgetDetails( {
	failedAudiences,
	showRetryButton,
	onRetry,
	onRequestAccess,
	errors,
	hasInsufficientPermissionsError,
} ) {
	if ( failedAudiences?.length ) {
		return (
			<div>
				<P>
					{ __(
						'Failed to create the following audiences:',
						'google-site-kit'
					) }
				</P>
				<ul className="googlesitekit-widget-audience-segmentation-error__failed-audiences">
					{ failedAudiences.map( ( audience ) => (
						<li key={ audience }>{ audience }</li>
					) ) }
				</ul>
				<Button onClick={ onRetry } danger>
					{ __( 'Retry', 'google-site-kit' ) }
				</Button>
			</div>
		);
	}

	if ( showRetryButton && onRetry ) {
		return (
			<Button onClick={ onRetry } danger>
				{ __( 'Retry', 'google-site-kit' ) }
			</Button>
		);
	}

	return (
		<ReportErrorActions
			moduleSlug="analytics-4"
			error={ errors }
			GetHelpLink={
				hasInsufficientPermissionsError ? GetHelpLink : undefined
			}
			hideGetHelpLink={ ! hasInsufficientPermissionsError }
			buttonVariant="danger"
			getHelpClassName="googlesitekit-error-retry-text"
			onRetry={ onRetry }
			onRequestAccess={ onRequestAccess }
		/>
	);
}

ErrorWidgetDetails.propTypes = {
	errors: PropTypes.arrayOf( PropTypes.object ).isRequired,
	onRetry: PropTypes.func.isRequired,
	onRequestAccess: PropTypes.func.isRequired,
	showRetryButton: PropTypes.bool,
	failedAudiences: PropTypes.arrayOf( PropTypes.string ),
	hasInsufficientPermissionsError: PropTypes.bool,
};
