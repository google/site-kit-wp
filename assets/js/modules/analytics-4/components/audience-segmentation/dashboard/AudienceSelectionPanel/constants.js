/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
