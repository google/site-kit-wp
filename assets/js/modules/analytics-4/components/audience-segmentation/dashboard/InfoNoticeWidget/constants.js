/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
