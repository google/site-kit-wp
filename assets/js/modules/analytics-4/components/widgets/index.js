/**
 * GA4 Key Metrics widget components.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as ConnectGA4CTATileWidget } from './ConnectGA4CTATileWidget';
export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
export { default as EngagedTrafficSourceWidget } from './EngagedTrafficSourceWidget';
export { default as LeastEngagingPagesWidget } from './LeastEngagingPagesWidget';
export { default as MostEngagingPagesWidget } from './MostEngagingPagesWidget';
export { default as NewVisitorsWidget } from './NewVisitorsWidget';
export { default as PopularContentWidget } from './PopularContentWidget';
export { default as PopularProductsWidget } from './PopularProductsWidget';
export { default as ReturningVisitorsWidget } from './ReturningVisitorsWidget';
export { default as TopCitiesWidget } from './TopCitiesWidget';
export { default as TopCitiesDrivingLeadsWidget } from './TopCitiesDrivingLeadsWidget';
export { default as TopCitiesDrivingAddToCartWidget } from './TopCitiesDrivingAddToCartWidget';
export { default as TopCitiesDrivingPurchasesWidget } from './TopCitiesDrivingPurchasesWidget';
export { default as TopDeviceDrivingPurchasesWidget } from './TopDeviceDrivingPurchasesWidget';
export { default as TopCountriesWidget } from './TopCountriesWidget';
export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
export { default as TopTrafficSourceDrivingAddToCartWidget } from './TopTrafficSourceDrivingAddToCartWidget';
export { default as TopTrafficSourceDrivingLeadsWidget } from './TopTrafficSourceDrivingLeadsWidget';
export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
export { default as TopTrafficSourceDrivingPurchasesWidget } from './TopTrafficSourceDrivingPurchasesWidget';
export { default as PagesPerVisitWidget } from './PagesPerVisitWidget';
export { default as VisitLengthWidget } from './VisitLengthWidget';
export { default as TopReturningVisitorPages } from './TopReturningVisitorPages';
export { default as VisitsPerVisitorWidget } from './VisitsPerVisitorWidget';
export { default as TopRecentTrendingPagesWidget } from './TopRecentTrendingPagesWidget';
export { default as TopCategoriesWidget } from './TopCategoriesWidget';
export { default as TopPagesDrivingLeadsWidget } from './TopPagesDrivingLeadsWidget';
export { default as PopularAuthorsWidget } from './PopularAuthorsWidget';
