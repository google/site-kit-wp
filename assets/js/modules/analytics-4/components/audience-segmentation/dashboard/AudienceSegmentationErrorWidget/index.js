/**
 * AudienceSegmentationErrorWidget component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { castArray } from 'lodash';
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useEffect } from '@wordpress/element';

/**
 * Internal dependencies
 */
import whenActive from '../../../../../../util/when-active';
import { useDispatch } from 'googlesitekit-data';
import { isInsufficientPermissionsError } from '../../../../../../util/errors';
import { CORE_UI } from '../../../../../../googlesitekit/datastore/ui/constants';
import { AUDIENCE_INFO_NOTICE_HIDE_UI } from '../InfoNoticeWidget/constants';
import ErrorWidgetContent from './ErrorWidgetContent';
import withIntersectionObserver from '../../../../../../util/withIntersectionObserver';
import { trackEvent } from '../../../../../../util';
import useViewContext from '../../../../../../hooks/useViewContext';

const ErrorWidgetContentWithIntersectionObserver =
	withIntersectionObserver( ErrorWidgetContent );

function AudienceSegmentationErrorWidget( {
	Widget,
	errors,
	onRetry,
	showRetryButton,
} ) {
	const viewContext = useViewContext();

	const { setValue } = useDispatch( CORE_UI );

	const errorsArray = errors ? castArray( errors ) : [];

	const hasInsufficientPermissionsError = errorsArray.some(
		isInsufficientPermissionsError
	);

	const handleRetry = () => {
		trackEvent(
			`${ viewContext }_audiences-all-tiles`,
			'data_loading_error_retry'
		).finally( () => {
			setValue( AUDIENCE_INFO_NOTICE_HIDE_UI, false );
			onRetry?.();
		} );
	};

	useEffect( () => {
		// Set UI key to hide the info notice.
		setValue( AUDIENCE_INFO_NOTICE_HIDE_UI, true );
	}, [ setValue ] );

	return (
		<ErrorWidgetContentWithIntersectionObserver
			Widget={ Widget }
			errors={ errorsArray }
			onRetry={ handleRetry }
			onRequestAccess={ () => {
				trackEvent(
					`${ viewContext }_audiences-all-tiles`,
					'insufficient_permissions_error_request_access'
				);
			} }
			showRetryButton={ showRetryButton }
			onInView={ () => {
				const action = hasInsufficientPermissionsError
					? 'insufficient_permissions_error'
					: 'data_loading_error';

				trackEvent( `${ viewContext }_audiences-all-tiles`, action );
			} }
		/>
	);
}

AudienceSegmentationErrorWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	errors: PropTypes.oneOfType( [
		PropTypes.object,
		PropTypes.arrayOf( PropTypes.object ),
	] ).isRequired,
	onRetry: PropTypes.func,
	showRetryButton: PropTypes.bool,
};

export default whenActive( { moduleName: 'analytics-4' } )(
	AudienceSegmentationErrorWidget
);
