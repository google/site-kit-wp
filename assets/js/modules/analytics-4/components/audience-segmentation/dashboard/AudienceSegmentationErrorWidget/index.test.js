/**
 * AudienceSegmentationErrorWidget tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { useIntersection as mockUseIntersection } from 'react-use';

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	provideModules,
	provideModuleRegistrations,
	provideUserInfo,
	fireEvent,
	waitForDefaultTimeouts,
	act,
} from '../../../../../../../../tests/js/test-utils';
import { withWidgetComponentProps } from '../../../../../../googlesitekit/widgets/util';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../../../../../googlesitekit/constants';
import { MODULES_ANALYTICS_4 } from '../../../../datastore/constants';
import AudienceSegmentationErrorWidget from '.';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../../../../../util/errors';
import * as tracking from '../../../../../../util/tracking';

jest.mock( 'react-use', () => ( {
	...jest.requireActual( 'react-use' ),
	useIntersection: jest.fn(),
} ) );

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'AudienceSegmentationErrorWidget', () => {
	let registry;

	beforeEach( () => {
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: false,
			intersectionRatio: 0,
		} ) );

		registry = createTestRegistry();
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );
		provideModuleRegistrations( registry );
		provideUserInfo( registry );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );
	} );

	afterEach( () => {
		mockTrackEvent.mockClear();
	} );

	const WidgetWithComponentProps = withWidgetComponentProps(
		'audienceSegmentationErrorWidget'
	)( AudienceSegmentationErrorWidget );

	describe( 'default error state', () => {
		let container,
			getByText,
			getByRole,
			queryByText,
			rerender,
			waitForRegistry;

		beforeEach( async () => {
			await registry.dispatch( MODULES_ANALYTICS_4 ).receiveError(
				{
					code: 'test-error-code',
					message: 'Test error message',
					data: {
						reason: '',
					},
				},
				'getReport',
				[
					{
						metrics: [
							{
								name: 'totalUsers',
							},
						],
						dimensions: [
							{
								name: 'date',
							},
						],
						startDate: '2020-08-11',
						endDate: '2020-09-07',
					},
				]
			);

			const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

			( {
				container,
				getByText,
				getByRole,
				queryByText,
				rerender,
				waitForRegistry,
			} = render( <WidgetWithComponentProps errors={ errors } />, {
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			} ) );

			await waitForRegistry();
		} );

		it( 'should render correctly', () => {
			expect( container ).toMatchSnapshot();

			expect(
				getByText( 'Your visitor groups data loading failed' )
			).toBeInTheDocument();

			expect(
				getByRole( 'button', { name: /retry/i } )
			).toBeInTheDocument();

			// Verify that it's not an "Insufficient permissions" error.
			expect(
				queryByText( 'Insufficient permissions' )
			).not.toBeInTheDocument();
			expect( queryByText( /request access/i ) ).not.toBeInTheDocument();
		} );

		it( 'should track an event when the widget is viewed', () => {
			const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps errors={ errors } /> );

			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-all-tiles',
				'data_loading_error'
			);
		} );

		it( 'should track an event when the "Retry" button is clicked', async () => {
			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			fireEvent.click( getByRole( 'button', { name: /retry/i } ) );

			// Allow the `trackEvent()` promise to resolve.
			await act( waitForDefaultTimeouts );

			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-all-tiles',
				'data_loading_error_retry'
			);
		} );
	} );

	describe( 'insufficient permissions error state', () => {
		let container,
			getByText,
			getByRole,
			queryByText,
			rerender,
			waitForRegistry;

		beforeEach( async () => {
			const [ accountID, propertyID, measurementID, webDataStreamID ] = [
				'12345',
				'34567',
				'56789',
				'78901',
			];

			await registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setAccountID( accountID );
			await registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setPropertyID( propertyID );
			await registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setMeasurementID( measurementID );
			await registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setWebDataStreamID( webDataStreamID );
			await registry.dispatch( MODULES_ANALYTICS_4 ).receiveError(
				{
					code: 'test-error-code',
					message: 'Test error message',
					data: {
						reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS,
					},
				},
				'getAccountID'
			);

			const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

			( {
				container,
				getByText,
				getByRole,
				queryByText,
				rerender,
				waitForRegistry,
			} = render( <WidgetWithComponentProps errors={ errors } />, {
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			} ) );

			await waitForRegistry();
		} );

		it( 'should render correctly', () => {
			expect( container ).toMatchSnapshot();

			expect(
				getByText( 'Insufficient permissions' )
			).toBeInTheDocument();

			expect(
				getByText(
					'Contact your administrator. Trouble getting access?'
				)
			).toBeInTheDocument();

			expect(
				getByRole( 'button', { name: /request access/i } )
			).toBeInTheDocument();

			// Verify that it's not a default error.
			expect(
				queryByText( 'Your visitor groups data loading failed' )
			).not.toBeInTheDocument();
			expect( queryByText( /retry/i ) ).not.toBeInTheDocument();
		} );

		it( 'should track an event when the widget is viewed', () => {
			const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps errors={ errors } /> );

			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-all-tiles',
				'insufficient_permissions_error'
			);
		} );

		it( 'should track an event when the "Request access" button is clicked', async () => {
			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			fireEvent.click(
				getByRole( 'button', { name: /request access/i } )
			);

			// Allow the `trackEvent()` promise to resolve.
			await act( waitForDefaultTimeouts );

			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-all-tiles',
				'insufficient_permissions_error_request_access'
			);
		} );
	} );

	it( 'should render a retry button when `onRetry` and `showRetryButton` props are passed', async () => {
		await registry.dispatch( MODULES_ANALYTICS_4 ).receiveError(
			{
				code: 'test-error-code',
				message: 'Test error message',
			},
			'getAccountID'
		);

		const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

		const handleRetrySpy = jest.fn();

		const { container, getByText, getByRole, waitForRegistry } = render(
			<WidgetWithComponentProps
				errors={ errors }
				onRetry={ handleRetrySpy }
				showRetryButton
			/>,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect(
			getByText( 'Your visitor groups data loading failed' )
		).toBeInTheDocument();

		expect( getByRole( 'button', { name: /retry/i } ) ).toBeInTheDocument();

		fireEvent.click( getByRole( 'button', { name: /retry/i } ) );

		// Allow the `trackEvent()` promise to resolve.
		await waitForDefaultTimeouts();

		expect( handleRetrySpy ).toHaveBeenCalledTimes( 1 );
	} );
} );
