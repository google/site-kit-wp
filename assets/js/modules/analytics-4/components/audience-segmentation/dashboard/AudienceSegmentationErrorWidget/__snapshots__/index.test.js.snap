// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceSegmentationErrorWidget should render a retry button when \`onRetry\` and \`showRetryButton\` props are passed 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--audienceSegmentationErrorWidget googlesitekit-widget--no-padding googlesitekit-audience-segmentation-error-widget"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid googlesitekit-audience-segmentation-error__widget-primary-cell mdc-layout-grid--collapsed"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-7-desktop mdc-layout-grid__cell--span-8-tablet"
          >
            <h3
              class="googlesitekit-publisher-win__title"
            >
              Your visitor groups data loading failed
            </h3>
            <div
              class="googlesitekit-widget-audience-segmentation-error__actions"
            >
              <button
                class="mdc-button mdc-button--raised mdc-button--danger"
                target="_self"
              >
                <span
                  class="mdc-button__label"
                >
                  Retry
                </span>
              </button>
            </div>
          </div>
          <div
            class="googlesitekit-widget-audience-segmentation-error__svg-wrapper mdc-layout-grid__cell"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceSegmentationErrorWidget default error state should render correctly 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--audienceSegmentationErrorWidget googlesitekit-widget--no-padding googlesitekit-audience-segmentation-error-widget"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid googlesitekit-audience-segmentation-error__widget-primary-cell mdc-layout-grid--collapsed"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-7-desktop mdc-layout-grid__cell--span-8-tablet"
          >
            <h3
              class="googlesitekit-publisher-win__title"
            >
              Your visitor groups data loading failed
            </h3>
            <div
              class="googlesitekit-widget-audience-segmentation-error__actions"
            >
              <div
                class="googlesitekit-report-error-actions"
              >
                <button
                  class="mdc-button mdc-button--raised mdc-button--danger"
                  target="_self"
                >
                  <span
                    class="mdc-button__label"
                  >
                    Retry
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-widget-audience-segmentation-error__svg-wrapper mdc-layout-grid__cell"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceSegmentationErrorWidget insufficient permissions error state should render correctly 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--audienceSegmentationErrorWidget googlesitekit-widget--no-padding googlesitekit-audience-segmentation-error-widget"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid googlesitekit-audience-segmentation-error__widget-primary-cell mdc-layout-grid--collapsed"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-7-desktop mdc-layout-grid__cell--span-8-tablet"
          >
            <h3
              class="googlesitekit-publisher-win__title"
            >
              Insufficient permissions
            </h3>
            <div
              class="googlesitekit-widget-audience-segmentation-error__actions"
            >
              <div
                class="googlesitekit-report-error-actions"
              >
                <a
                  aria-label="Request access (opens in a new tab)"
                  class="mdc-button mdc-button--raised mdc-button--danger"
                  href="https://accounts.google.com/accountchooser?continue=https%3A%2F%2Fanalytics.google.com%2Fanalytics%2Fweb%2F%23%2Fa12345p34567%2Fadmin%2Fstreams%2Ftable%2F78901&Email=wapuu.wordpress%40gmail.com"
                  role="button"
                  target="_blank"
                >
                  <span
                    class="mdc-button__label"
                  >
                    Request access
                  </span>
                </a>
                <div
                  class="googlesitekit-error-retry-text"
                >
                  Contact your administrator. Trouble getting access? 
                  <a
                    aria-label="Get help (opens in a new tab)"
                    class="googlesitekit-cta-link"
                    href="undefined?error_id=analytics-4_insufficient_permissions"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <span
                      class="googlesitekit-cta-link__contents"
                    >
                      Get help
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-widget-audience-segmentation-error__svg-wrapper mdc-layout-grid__cell"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
